TodoApp/TodoApp.Api.Tests/Controllers/ItemListController.test.cs

```diff
@@ -0,0 +1,130 @@
+            var actualStatusCode = await ResolveResult(result);
+            message.TryGetContentValue(out TResult actualItems);
+            return (actualStatusCode, actualItems);
+        }
+
+        public static async Task<HttpStatusCode> ResolveResult<TController>(this TController controller, Func<TController, Task<IHttpActionResult>> actionSelector)
+        {
+            var result = await actionSelector(controller);
+            var actualStatusCode = await ResolveResult(result);
+            return actualStatusCode;
+        }
+    }
+
+    [TestFixture]
+    public class ItemListControllerTest
+    {
+        private ItemListController _controller;
+        
+        [SetUp]
+        public void SetUp()
+        {
+            _controller = new ItemListController
+            {
+                Request = new HttpRequestMessage(),
+                Configuration = new HttpConfiguration()
+            };
+        }
+
+        [Test]
+        public async Task GetAllItems_AllItemsReturned()
+        {
+            var expectedItems = new[]
+            {
+                new ItemModel {Id = Guid.Parse("c5cc89a0-ab8d-4328-9000-3da679ec02d3"), Text = "Make a cofee"},
+                new ItemModel {Id = Guid.Parse("55b0d56d-48d7-4f93-bd73-e4b801e26faa"), Text = "Make second coffee"},
+                new ItemModel {Id = Guid.Parse("c5cc89a0-ab8d-4328-9000-3da679ec02d2"), Text = "Add some coffee"},
+                new ItemModel {Id = Guid.Parse("250be0cc-438e-46cc-a0fe-549f4d3409e2"), Text = "Coffee overflow"}
+            };
+
+            var (actualStatusCode, actualItems) =
+                await _controller.ResolveResult<ItemListController, ItemModel[]>(controller => controller.GetAllAsync());
+
+            Assert.That(actualStatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(actualItems, Is.EqualTo(expectedItems).UsingItemModelComparer());
+        }
+
+        [Test]
+        public async Task GetItem_ExistingId_ItemReturned()
+        {
+            var expectedItem = new ItemModel {Id = Guid.Parse("c5cc89a0-ab8d-4328-9000-3da679ec02d3"), Text = "Make a cofee"};
+            var guid = Guid.Parse("c5cc89a0-ab8d-4328-9000-3da679ec02d3");
+
+            var (actualStatusCode, actualItem) = await _controller.ResolveResult<ItemListController, ItemModel>(controller => controller.GetAsync(guid));
+
+            Assert.That(actualStatusCode, Is.EqualTo(HttpStatusCode.OK));
+            Assert.That(actualItem, Is.EqualTo(expectedItem).UsingItemModelComparer());
+        }
+
+        [Test]
+        public async Task PostNewItem_UniqueItem_ItemAdded()
+        {
+            var expectedItem = new ItemModel { Id = Guid.Parse("a09c0705-b162-4443-b497-9812e6b5c5aa"), Text = "Another coffee" };
+
+            var actualStatusCode = await _controller.ResolveResult(controller => controller.PostAsync(expectedItem));
+            
+            Assert.That(actualStatusCode, Is.EqualTo(HttpStatusCode.Created));
```

Comment:
You need to test whether correct item and location were returned here - it will be a bit tricky one with the extensions you already have.

You could even decide to replace tuples with a "regular" `struct` or `class` of yours to provide good readability to your tests or reinvent the extensions from the ground. We might actually discuss this further in person all-together with generics in the before mentioned extension methods and, hopefully, come up with some event funnier and more inspiring solution. However, before that, go ahead and try to solve this puzzle on your own, please 😃